### PR TITLE
Add re-implementation of the build command

### LINKED ISSUE #4868: Provide a different way of building images with minikube
Currently we delegate all building of images to `docker`, using `minikube docker-env`.

This requires the user to install Docker on their machine, and then learn how to set it up...

https://kubernetes.io/docs/tutorials/hello-minikube/

> [minikube/Dockerfile](https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/minikube/Dockerfile)
```Dockerfile
FROM node:6.14.2
EXPOSE 8080
COPY server.js .
CMD node server.js
```

> [minikube/server.js ](https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/minikube/server.js)
```js
var http = require('http');

var handleRequest = function(request, response) {
  console.log('Received request for URL: ' + request.url);
  response.writeHead(200);
  response.end('Hello World!');
};
var www = http.createServer(handleRequest);
www.listen(8080);
```

> For more information on the docker build command, read the [Docker documentation](https://docs.docker.com/engine/reference/commandline/build/).

If the user doesn't already have a local installation of docker, they can't build the image!

----

We could do better, by providing an abstraction that will simply do the build for them:

```
$ minikube build -- -t hello-node minikube
💾  Downloading docker 18.09.8
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM node:6.14.2
6.14.2: Pulling from library/node
3d77ce4481b1: Pull complete 
7d2f32934963: Pull complete 
0c5cf711b890: Pull complete 
9593dc852d6b: Pull complete 
4e3b8a1eb914: Pull complete 
ddcf13cc1951: Pull complete 
2e460d114172: Pull complete 
d94b1226fbf2: Pull complete 
Digest: sha256:62b9d88be259a344eb0b4e0dd1b12347acfe41c1bb0f84c3980262f8032acc5a
Status: Downloaded newer image for node:6.14.2
 ---> 00165cd5d0c0
Step 2/4 : EXPOSE 8080
 ---> Running in 2a302085e433
Removing intermediate container 2a302085e433
 ---> 9172f65af846
Step 3/4 : COPY server.js .
 ---> 035625e5e23f
Step 4/4 : CMD node server.js
 ---> Running in 1771091ed23a
Removing intermediate container 1771091ed23a
 ---> 250208286ec5
Successfully built 250208286ec5
Successfully tagged hello-node:latest
```
Then the image is built right on the VM, and ready to be used from the minikube pods:

```
$ minikube ssh sudo crictl images hello-node
IMAGE               TAG                 IMAGE ID            SIZE
hello-node          latest              250208286ec58       660MB
$ minikube kubectl -- create deployment hello-node --image=hello-node
💾  Downloading kubectl v1.15.0
deployment.apps/hello-node created
```

As usual have to edit the pull policy, when not using a registry but the local images.

``` yaml
      containers:
      - image: hello-node:latest
        imagePullPolicy: IfNotPresent
        name: hello-node
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
```

Change `Always`, as per https://kubernetes.io/docs/concepts/containers/images/

----

Eventually we could improve this by **not** using the Docker daemon but e.g. `buildah` :

* https://github.com/containers/libpod/blob/master/docs/podman-build.1.md

* https://github.com/containers/buildah/blob/master/docs/buildah-bud.md

That way the user don't have to have the dockerd running, but can use containerd or cri-o.

----

This project could also be interesting, eventually:
https://github.com/GoogleContainerTools/kaniko

That is: building the images in Kubernetes instead ?
With enough kernel support, also doable with buildah.

https://opensource.com/article/19/3/tips-tricks-rootless-buildah

### PR DESCRIPTION
**Pull request by afbjorklund**
This is a rebase of PR #10742

Closes #4868

### CODE REVIEW

**Review by medyagh — COMMENTED** [review]
since this is a new feature @afbjorklund would you please add an integration test ? maybe in FunctionalTest Next To docker-env test?

that way it could serve as both Example of using and also we will ensure we never break it

**Comment by afbjorklund** [comment]
It seems that there were no tests for any of the previous `minikube image` commands ?

**Comment by afbjorklund** [comment]
> It seems that there were no tests for any of the previous `minikube image` commands ?

Obviously I only looked for `image ls`, there were tests for `image load` and `image rm`.

**Review by medyagh — CHANGES_REQUESTED** [review]
windows integration test failing 
```
2021-04-23T16:06:32.9069971Z === CONT  TestFunctional/parallel/BuildImage
2021-04-23T16:06:32.9074954Z     functional_test.go:291: (dbg) Non-zero exit: ./minikube-windows-amd64.exe ssh -p functional-20210423160039-16840 -- docker image inspect my-image: exit status 1 (1.591562s)
2021-04-23T16:06:32.9093489Z         
2021-04-23T16:06:32.9095921Z         -- stdout --
2021-04-23T16:06:32.9103442Z         	[]
2021-04-23T16:06:32.9105356Z         	Error: No such image: my-image
2021-04-23T16:06:32.9106462Z         
2021-04-23T16:06:32.9110386Z         -- /stdout --
2021-04-23T16:06:32.9112516Z         ** stderr ** 
2021-04-23T16:06:32.9116138Z         	ssh: Process exited with status 1
2021-04-23T16:06:32.9122414Z         
2021-04-23T16:06:32.9124982Z         ** /stderr **
2021-04-23T16:06:32.9131411Z     functional_test.go:329: listing images: exit status 1
2021-04-23T16:06:32.9133450Z         
2021-04-23T16:06:32.9138610Z         -- stdout --
2021-04-23T16:06:32.9142331Z         	[]
2021-04-23T16:06:32.9146494Z         	Error: No such image: my-image
2021-04-23T16:06:32.9149756Z         
2021-04-23T16:06:32.9160974Z         -- /stdout --
2021-04-23T16:06:32.9165037Z         ** stderr ** 
2021-04-23T16:06:32.9166630Z         	ssh: Process exited with status 1
2021-04-23T16:06:32.9168890Z         
2021-04-23T16:06:32.9172849Z         ** /stderr **
2021-04-23T16:06:32.9177731Z     helpers_test.go:218: -----------------------post-mortem--------------------------------
```

File: go.mod [hunk 1]
```diff
@@ -21,7 +21,7 @@ require (
 	github.com/cloudfoundry-attic/jibber_jabber v0.0.0-20151120183258-bcc4c8345a21
 	github.com/cloudfoundry/jibber_jabber v0.0.0-20151120183258-bcc4c8345a21 // indirect
 	github.com/docker/cli v0.0.0-20200303162255-7d407207c304 // indirect
-	github.com/docker/docker v17.12.0-ce-rc1.0.20200916142827-bd33bbf0497b+incompatible
```
Comment (medyagh) [inline go.mod:24]: is there a PR upstream that we could comment here to remove the replace after the PR gets merged ?

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/minikube/cruntime/containerd.go [hunk 2]
```diff
@@ -277,6 +279,109 @@ func (r *Containerd) RemoveImage(name string) error {
...
+	}
+	tmp := strings.TrimSpace(rr.Stdout.String())
+	cmd := exec.Command("git", "clone", src, tmp)
+	if _, err := cr.RunCmd(cmd); err != nil {
+		return "", err
+	}
+	return tmp, nil
+}
+
+func downloadRemote(cr CommandRunner, src string) (string, error) {
```
Comment (medyagh) [inline pkg/minikube/cruntime/containerd.go:296]: add unit test for this ? or maybe break down the logic of detecting remote and add unit tests, so we know exactly what user would input

**Comment by afbjorklund** [comment]
The integration test is failing with containerd/buildkitd it seems, due to the naming of the image:
If you build something with a short name ("myimage"), it gets tagged as "docker.io/library/myimage"

```
        #8 exporting to image
        #8 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
        #8 exporting layers
        #8 exporting layers 0.2s done
        #8 exporting manifest sha256:d7a819d73f2ab7530f6008982101303db287581998611fc3986ad9ace3ff9089 done
        #8 exporting config sha256:a8fdd9c45e461cee2130ba01b0007d5c08c9058b429fd26d8d0bdaa42d3eb2c8 done
        #8 naming to my-image:latest done
        #8 DONE 0.2s
```

```
$ sudo crictl images my-image
IMAGE                                     TAG                  IMAGE ID            SIZE
docker.io/library/my-image                latest               a8fdd9c45e461       767kB
```

**Review by medyagh — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (afbjorklund) [reply go.mod:24]: Alternatively one could upgrade everything to the Docker 20.10 API, but that is much more intrusive and big

File: pkg/minikube/cruntime/containerd.go (hunk 2 above)
Comment (afbjorklund) [reply pkg/minikube/cruntime/containerd.go:296]: Right, but then we need to start a web server or something to host the files to build (the ones in testdata/build) ?

File: test/integration/functional_test.go [hunk 3]
```diff
@@ -294,6 +295,54 @@ func inspectImage(ctx context.Context, t *testing.T, profile string, image strin
...
+		t.Logf("(dbg) Stdout: %s:\n%s", rr.Command(), rr.Stdout)
+	}
+	if rr.Stderr.Len() > 0 {
+		t.Logf("(dbg) Stderr: %s:\n%s", rr.Command(), rr.Stderr)
+	}
+
+	// make sure the image was correctly built
+	rr, err = inspectImage(ctx, t, profile, newImage)
+	if err != nil {
+		t.Fatalf("listing images: %v\n%s", err, rr.Output())
```
Comment (medyagh) [inline test/integration/functional_test.go:329]: maybe if fatal, list all images in the cluster for debugging..?

**Review by medyagh — CHANGES_REQUESTED** [review]

**Comment by afbjorklund** [comment]
> windows integration test failing

It's related to the DOS paths from the host:

```
=== CONT  TestFunctional/parallel/BuildImage
    functional_test.go:315: (dbg) Done: ./minikube-windows-amd64.exe -p functional-20210423160130-2736 image build -t my-image testdata\build: (10.8477205s)
    functional_test.go:323: (dbg) Stderr: ./minikube-windows-amd64.exe -p functional-20210423160130-2736 image build -t my-image testdata\build:
        unable to prepare context: path "C:\\Users\\jenkins\\AppData\\Local\\Temp\\build.211045959.tar" not found
    functional_test.go:291: (dbg) Run:  ./minikube-windows-amd64.exe ssh -p functional-20210423160130-2736 -- docker image inspect my-image
```

The code needs to be more careful about which paths are from the host (possibly weird), and which are on machine (and sane)

File: pkg/minikube/cruntime/containerd.go (hunk 2 above)
Comment (medyagh) [reply pkg/minikube/cruntime/containerd.go:296]: I think the detection part of the (the scheme detection) could be put in a separate func with unit test ?

Comment (afbjorklund) [reply pkg/minikube/cruntime/containerd.go:296]: We only implement it for containerd, both docker and podman have their own (internal) implementation...

File: go.mod (hunk 1 above)
Comment (afbjorklund) [reply go.mod:24]: The patch was: https://github.com/moby/moby/commit/c3a0a3744636069f43197eb18245aaae89f568e5

**Review by medyagh — APPROVED** [review]
Looks good to me @ , Thank you for this feature. I am excited to see how our users will use this

**Pull request merged**